src/game/engine/internals/Transform.js

```diff
@@ -0,0 +1,61 @@
+import { Vector } from '../../shared';
+
+class Transform {
+  #position = Vector.Zero;
+
+  #origin = Vector.Zero;
+
+  #width = 0;
+
+  #height = 0;
+
+  get width() {
+    return this.#width;
+  }
+
+  /**
+   * @param {number} w
+   */
+  set width(w) {
+    this.#width = w;
+  }
+
+  get height() {
+    return this.#height;
+  }
+
+  /**
+   * @param {number} h
+   */
+  set height(h) {
+    this.#height = h;
+  }
+
+  get position() {
+    return this.#position;
+  }
+
+  /**
+   * @param {import('../../shared').Vector} v
+   */
+  set position(v) {
+    this.#position = v;
+  }
+
+  get origin() {
+    return this.#origin;
+  }
+
+  /**
+   * @param {import('../../shared').Vector} v
+   */
+  updatePosition(v) {
+    this.#position = this.#position.add(v);
+  }
+
+  updateOrigin() {
+    this.#origin = new Vector(this.#position.x + 0.5 * this.#width, this.#position.y + 0.5 * this.#height);
```

Comment:
Ja bym zamienił kolejność, tzn. `origin` ustawił jako główny wektor do środka obiektu, a `position` dla kompatybilności ze `sprite` (które teraz wykorzystujemy) wyliczył tak:
```js
this.#position= new Vector(this.#origin.x - 0.5 * this.#width, this.#origin.y - 0.5 * this.#height);
```
Możliwe, że w przyszłości `position` w ogóle nam nie będzie potrzebne i to przeliczanie przerzuci się do draw `sprite`.

Reply:
Bawiłam się trochę tym, zrobiłam wszystko zgodnie ze wskazówkami, ale nie działało tak, jak tego oczekujemy - collidery ptaków nie zmieniały swojej pozycji (cały czas lewy górny róg). Collider **Trexa** też ustawił się w (0, 0). Nie wiem z czego to może wynikać. Gdy usunęłam ze sceny Trexa (inaczej od razu po pojawieniu się ptaka był koniec gry), to ptak poruszał się normalnie, więc tylko coś z tymi colliderami nie gra.

Wrzucam kod `Transform.js` po zmianach, które zasugerowałeś:

```javascript
import { Vector } from '../../shared';

class Transform {
  #position = Vector.Zero;

  #origin = Vector.Zero;

  #width = 0;

  #height = 0;

  get width() {
    return this.#width;
  }

  /**
   * @param {number} w
   */
  set width(w) {
    this.#width = w;
  }

  get height() {
    return this.#height;
  }

  /**
   * @param {number} h
   */
  set height(h) {
    this.#height = h;
  }

  get position() {
    return this.#position;
  }

  /**
   * @param {import('../../shared').Vector} v
   */
  set position(v) {
    this.#position = v;
  }

  get origin() {
    return this.#origin;
  }

  /**
   * @param {import('../../shared').Vector} v
   */
  set origin(v) {
    this.#origin = v;
    this.#updatePosition();
  }

  #updatePosition() {
    this.#position = new Vector(this.#origin.x - 0.5 * this.#width, this.#origin.y - 0.5 * this.#height);
  }
}

export { Transform };
```

w `Rigidbody.js` 
```javascript
  update(transform) {
    ...
    transform.origin = transform.origin.add(this.#resultVector);
  }
```
Może wypatrzysz jakiś błąd.

Reply:
Okej, nvm,  już chyba widzę

Reply:
Błąd polegał, jedynie na tym, że `Trex` oraz `Cactus` same manipulowały pozycją (a nie za pomocą `rigidbody`, tak jak chcemy), więc żeby być z nimi kompatybilni trzeba było dodać te dwie linijki:
```js
#updateOrigin() {
  this.#origin = new Vector(this.#position.x + 0.5 * this.#width, this.#position.y + 0.5 * this.#height);
}

set position(v) {
  this.#position = v;
  this.#updateOrigin();
}
```